server/service/src/stock_line/tests/historical_stock.rs

```diff
@@ -0,0 +1,479 @@
+#[cfg(test)]
+mod query {
+    use chrono::{NaiveDate, NaiveDateTime};
+    use repository::mock::{mock_item_a, mock_name_customer_a, mock_name_store_b, mock_store_a};
+    use repository::mock::{mock_user_account_a, MockDataInserts};
+    use repository::{
+        InvoiceLineRow, InvoiceLineType, InvoiceRow, InvoiceStatus, InvoiceType, StockLineRow,
+        StockLineRowRepository, Upsert,
+    };
+    use util::date_now;
+
+    use crate::service_provider::ServiceContext;
+    use crate::test_helpers::{setup_all_and_service_provider, ServiceTestContext};
+
+    static mut INVOICE_NUMBER: i64 = 0;
+
+    // These are all from default mock data
+    static ITEM_ID: &str = "item_a";
+    static STORE_ID: &str = "store_a";
+
+    static STOCK_LINE_A: &str = "stock_line_a";
+    static STOCK_LINE_B: &str = "stock_line_b";
+    static STOCK_LINE_C: &str = "stock_line_c";
+
+    fn get_midnight(year: i32, month: u32, day: u32) -> NaiveDateTime {
+        NaiveDate::from_ymd_opt(year, month, day)
+            .unwrap()
+            .and_hms_opt(0, 0, 0)
+            .unwrap()
+    }
+
+    fn get_midday(year: i32, month: u32, day: u32) -> NaiveDateTime {
+        NaiveDate::from_ymd_opt(year, month, day)
+            .unwrap()
+            .and_hms_opt(12, 0, 0)
+            .unwrap()
+    }
+
+    fn next_invoice_number() -> i64 {
+        unsafe {
+            INVOICE_NUMBER += 1;
+            INVOICE_NUMBER
+        }
+    }
+
+    fn update_stock(
+        ctx: &ServiceContext,
+        datetime: NaiveDateTime,
+        stock_line_id: String,
+        pack_size: f64,
+        number_of_packs: f64,
+        batch: String,
+    ) -> () {
+        let invoice_type = if number_of_packs > 0.0 {
+            InvoiceType::InboundShipment
+        } else {
+            InvoiceType::OutboundShipment
+        };
+        let invoice_line_type = if number_of_packs > 0.0 {
+            InvoiceLineType::StockIn
+        } else {
+            InvoiceLineType::StockOut
+        };
+        let name = if number_of_packs > 0.0 {
+            mock_name_store_b().id // Supplier
+        } else {
+            mock_name_customer_a().id // Customer
+        };
+
+        let old_stock_line = StockLineRowRepository::new(&ctx.connection)
+            .find_one_by_id(&stock_line_id)
+            .unwrap()
+            .unwrap_or_default();
+        let new_number_of_packs = old_stock_line.total_number_of_packs + number_of_packs;
+
+        let invoice_number = next_invoice_number();
```

Comment:
Btw for tests I wouldn't bother too much with invoice nubmers or other selecting correct supplier/customer (unless we using service), I think that woudl reduce the helper methods for tests, also would try to comment inline (just the SOH), so the data definition for test would be something like this:

```rust
        let stock_movements = vec![
            (100, StockIn, "A", "2021-01-01"),  // A, SOH = 100,
            (50, StockOut, "A", "2021-01-02"),  // A, SOH = 50
            (50, StockOut, "A", "2021-01-03"),  // A, SOH = 0
            (100, StockIn, "A", "2021-01-04"),  // A, SOH = 100
            (100, StockOut, "A", "2021-01-05"), // A, SOH = 0
        ];
```

Not a change request, just a thought for balancing out test helpers code vs clarity of data definition

Reply:
This is a good idea, would help with keeping comments in sync with code too...